cpp/src/strings/copy/repeat_join.cu

```diff
@@ -0,0 +1,158 @@
+
+  CUDF_EXPECTS(input.size() <= std::numeric_limits<size_type>::max() / repeat_times,
+               "The output string has size that exceeds the maximum allowed size.");
+
+  auto buff =
+    rmm::device_buffer(repeat_times * input.size(), stream, rmm::mr::get_current_device_resource());
+  for (size_type i = 0, str_size = input.size(); i < repeat_times; ++i) {
+    auto const offset = i * str_size;
+    thrust::copy(rmm::exec_policy(stream),
+                 input.data(),
+                 input.data() + str_size,
+                 static_cast<char*>(buff.data()) + offset);
+  }
+
+  return string_scalar(string_view(static_cast<char*>(buff.data()), buff.size()), true, stream, mr);
+}
+
+namespace {
+/**
+ * @brief Functor to compute string sizes and repeat the input strings.
+ */
+template <bool has_nulls>
+struct compute_size_and_repeat_fn {
+  column_device_view const strings_dv;
+  size_type repeat_times;
+
+  offset_type* d_offsets{nullptr};
+
+  // If d_chars == nullptr: only compute sizes of the output strings.
+  // If d_chars != nullptr: only repeat strings.
+  char* d_chars{nullptr};
+
+  // Set `1` or `0` for the validities of the output strings if `has_nulls` is `true`.
+  int8_t* d_validities{nullptr};
+
+  __device__ void operator()(size_type const idx) const noexcept
+  {
+    if (!d_chars) {
+      auto const is_valid = !has_nulls || strings_dv.is_valid_nocheck(idx);
+      d_offsets[idx]      = is_valid && repeat_times > 0
+                         ? repeat_times * strings_dv.element<string_view>(idx).size_bytes()
+                         : 0;
+      if constexpr (has_nulls) { d_validities[idx] = is_valid; }
+    } else if ((!has_nulls || d_validities[idx]) && repeat_times > 0) {
+      auto const d_str = strings_dv.element<string_view>(idx);
+      auto output_ptr  = d_chars + d_offsets[idx];
+      for (size_type i = 0; i < repeat_times; ++i) {
+        output_ptr = detail::copy_string(output_ptr, d_str);
+      }
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> repeat_join(strings_column_view const& input,
+                                    size_type repeat_times,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto const num_rows = input.size();
+  if (num_rows == 0) { return detail::make_empty_strings_column(stream, mr); }
+  if (num_rows == 1) { return std::make_unique<column>(input.parent(), stream, mr); }
+
+  if (repeat_times > 0) {
+    auto const size_start =
+      cudf::detail::get_value<size_type>(input.offsets(), input.offset(), stream);
+    auto const size_end =
+      cudf::detail::get_value<size_type>(input.offsets(), input.offset() + num_rows, stream);
+    CUDF_EXPECTS(size_end - size_start <= std::numeric_limits<size_type>::max() / repeat_times,
+                 "The output strings have total size that exceeds the maximum allowed size.");
+  }
```

Comment:
Can this check be done by the caller? This overhead seems a waste for a libcudf API.

Reply:
I think we should do the checking here. It's not very much expensive while we can make sure that we will not have subtle bugs hidden (if the output size is out of range, we may end up with an undefined result instead of an exception). The caller may not always remember to do that.

Reply:
libcudf is mainly intended for performance and not for convenience. This check can be burdensome for someone who has data that they know will never go out of bounds but must endure the performance hit of this check without choice.
cc @jrhemstad for his opinion.

Reply:
There is a similar situation: https://github.com/rapidsai/cudf/issues/8401
```
RuntimeError: cuDF failure at: ../src/copying/concatenate.cu:365: Total number of concatenated rows exceeds size_type range
```

Reply:
Just watched the DS All Hands today and saw that we may have the ability to support big strings column very soon. If we switch to use `int64_t` type for indexing then we will no longer have such error checking.

Reply:
https://github.com/rapidsai/cudf/issues/3958#issuecomment-803683107

Reply:
If not `int64_t` for indexing all columns then at least some other way to handle large strings column (`int64_t` for indexing only strings column?) would be good enough :(

Reply:
https://github.com/rapidsai/cudf/issues/5505

Reply:
Two things:

1. @ttnghia We already had a discussion about int64_t indexing. Lots of them and they were long.  There are a lot of issues with trying to support this so we are working on it a bit at a time.

2. Generally the checking happens when you know the size of the result you need to allocate. That way there is no extra overhead. You need to be careful because of the possibility of overflowing the int32_t size type and getting either illegal memory access errors or possibly data corruption. You should calculate the size as a int64_t and then compare it to the limits allowed for a column.

Reply:
Thanks, Bobby. Unfortunately, just checking the column size (reading from CPU memory) is not enough as the input column can be a sliced column. As such, we have to read the offsets (with stream sync) from device memory to compute the size of the column segment we are going to operate on. I'm going to remove the error check here to comply with cudf dev philosophy but we will definitely do the check somewhere else at the caller site (which may do a similar thing with similar overheadf: reading offsets from device memory with stream sync).

Reply:
I am not saying use the input column size. I am saying use the size of the output column that we are going to use to allocate data. In this case `make_strings_children` is the one that executes the code to get the size of the data to be allocated. The problem is that it is using an `int32_t` for the result of the computation, not an `int64_t`. In most cases that is fine because if it overflows it will overflow to be a negative value and then when you try to allocate a column with negative rows in it there is a check that will fail.

In this case we have a problem because I could allocate enough to overflow twice and make it positive again. This same thing happened in join with really fun consequences. That is because we are multiplying, not just adding, although even with adding it is still a possibility. 
`2 * 2147483647` results in `-2` for `int32_t`
`3 * 2147483647` results in `2147483645` for `int32_t`

If I have a string column that is already very close to 2 GiB in size and I ask to repeat it `3` times. I will allocate something very close to 2 GiB for the output because it overflowed twice, and then depending on how the kernel is written proceed to write 4 GiB of data after the end of the buffer.  To me that feels like a big enough issue that cudf should check it.  But that is just me.

What I would like to see happen is to have `make_strings_children` check this some how, but there is no super simple/efficient way to do it because `make_strings_children` is creating the offsets column using an exclusive scan and then pulling the final result back to see what the size is that it needs to allocate.